portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -94,17 +110,62 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
       response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage());
     }
 
-    User currentUser = userService.getCurrentUser();
-    String userEmail = currentUser.getEmail();
-    String userId = currentUser.getUserId();
-    datastore.put(createCommentEntity(newComment, userEmail, userId));
+    String imageUrl = getUploadedFileUrl(request, "image", "image/");
+    if (imageUrl != null || newComment != null) {
+      User currentUser = userService.getCurrentUser();
+      String userEmail = currentUser.getEmail();
+      String userId = currentUser.getUserId();
+      datastore.put(createCommentEntity(newComment, userEmail, userId, imageUrl));
+    }
 
-    // Redirect back to the HTML page.
+    // Redirect to the HTML page.
     response.sendRedirect("/index.html");
   }
 
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file of 
+   *  type contentType. 
+   */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName, String contentType) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    // getUploads returns Map with values as List of BlobKey for any files that were uploaded,
+    // keyed by the "name" field of the uploaded form.
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
```

Comment:
Ok so I am a bit confused here - when the user adds a comment, they also pick an image file and the same single POST contains both the comment & the binary data for the image?  Did I get that right?

Looks like blobstore can automatically find those uploads which is news to me, so just trying to make sure I follow the logic here :-)

Reply:
Yes, when the user adds a comment, they can also pick an image file.
The POST request first goes to Blobstore (blobstoreService.getUploads(request) contains only the image). Then Blobstore forwards the request to DataServlet(which gets the comment text using getParameter). 

I'm not sure if the comment text is processed in Blobstore, but my current understanding is that the comment is kept in the parameters of the request.

Reply:
I added the file upload input within the comment form to bind comments to images, but is it better to split the form?

Reply:
I think keep the form together (so that the comment and picture upload are one atomic action for the user), and as discussed in our meeting think about if there is a way to refactor your servlets + business logic.  You might need to look at the blobstore docs to see if there is a way to directly insert the image binary data from within you code.